src/native/http_connection_manager.c

```diff
@@ -0,0 +1,241 @@
+#include <aws/io/socket_channel_handler.h>
+#include <aws/io/tls_channel_handler.h>
+
+#include <aws/http/connection.h>
+#include <aws/http/connection_manager.h>
+#include <aws/http/http.h>
+
+/* on 32-bit platforms, casting pointers to longs throws a warning we don't need */
+#if UINTPTR_MAX == 0xffffffff
+#    if defined(_MSC_VER)
+#        pragma warning(push)
+#        pragma warning(disable : 4305) /* 'type cast': truncation from 'jlong' to 'jni_tls_ctx_options *' */
+#    else
+#        pragma GCC diagnostic push
+#        pragma GCC diagnostic ignored "-Wpointer-to-int-cast"
+#        pragma GCC diagnostic ignored "-Wint-to-pointer-cast"
+#    endif
+#endif
+
+/* methods of HttpConnectionPoolManager */
+static struct { jmethodID onConnectionAcquired; } s_http_connection_manager;
+
+void s_cache_http_conn_manager(JNIEnv *env) {
+    jclass cls = (*env)->FindClass(env, "software/amazon/awssdk/crt/http/HttpConnectionPoolManager");
+    AWS_FATAL_ASSERT(cls);
+
+    s_http_connection_manager.onConnectionAcquired = (*env)->GetMethodID(env, cls, "onConnectionAcquired", "(JI)V");
+    AWS_FATAL_ASSERT(s_http_connection_manager.onConnectionAcquired);
+}
+
+JNIEXPORT jlong JNICALL Java_software_amazon_awssdk_crt_http_HttpConnectionPoolManager_httpConnectionManagerNew(
+    JNIEnv *env,
+    jclass jni_class,
+    jlong jni_client_bootstrap,
+    jlong jni_socket_options,
+    jlong jni_tls_ctx,
+    jint jni_window_size,
+    jstring jni_endpoint,
+    jint jni_port,
+    jint jni_max_conns) {
+
+    (void)jni_class;
+
+    struct aws_client_bootstrap *client_bootstrap = (struct aws_client_bootstrap *)jni_client_bootstrap;
+    struct aws_socket_options *socket_options = (struct aws_socket_options *)jni_socket_options;
+    struct aws_tls_ctx *tls_ctx = (struct aws_tls_ctx *)jni_tls_ctx;
+
+    if (!client_bootstrap) {
+        aws_jni_throw_runtime_exception(env, "ClientBootstrap can't be null");
+        return (jlong)NULL;
+    }
+
+    if (!socket_options) {
+        aws_jni_throw_runtime_exception(env, "SocketOptions can't be null");
+        return (jlong)NULL;
+    }
+
+    struct aws_allocator *allocator = aws_jni_get_allocator();
+    struct aws_byte_cursor endpoint = aws_jni_byte_cursor_from_jstring(env, jni_endpoint);
+
+    if (jni_port <= 0 || 65535 < jni_port) {
+        aws_jni_throw_runtime_exception(env, "Port must be between 1 and 65535");
+        return (jlong)NULL;
+    }
+
+    if (jni_window_size <= 0) {
+        aws_jni_throw_runtime_exception(env, "Window Size must be > 0");
+        return (jlong)NULL;
+    }
+
+    if (jni_max_conns <= 0) {
+        aws_jni_throw_runtime_exception(env, "Max Connections must be > 0");
+        return (jlong)NULL;
+    }
+
+    uint16_t port = (uint16_t)jni_port;
+
+    int use_tls = (jni_tls_ctx != 0);
+
+    struct aws_tls_connection_options tls_conn_options = {0};
+
+    if (use_tls) {
+        aws_tls_connection_options_init_from_ctx(&tls_conn_options, tls_ctx);
```

Comment:
tls_ctx isn't cheap and can technically be reused, if mtls isn't being used. Do we need to create a new ctx each time?

Reply:
I just moved this code over from `http_connection.c` with very few changes. I don't think that change should block this PR